execution_engine/src/runtime/mod.rs

```diff
@@ -3223,4 +3242,72 @@ where
             None => Err(Error::InvalidContract(contract_hash)),
         }
     }
+
+    fn add_message_topic(&mut self, topic_name: String) -> Result<Result<(), ApiError>, Error> {
+        let topic_hash = crypto::blake2b(AsRef::<[u8]>::as_ref(&topic_name)).into();
+
+        self.context
+            .add_message_topic(topic_name, topic_hash)
+            .map(|ret| ret.map_err(ApiError::from))
+    }
+
+    fn emit_message(
+        &mut self,
+        topic_name: String,
+        message: MessagePayload,
+    ) -> Result<Result<(), ApiError>, Trap> {
+        let entity_addr = self
+            .context
+            .get_entity_address()
+            .into_hash()
+            .ok_or(Error::InvalidContext)?;
+
+        let topic_name_hash = crypto::blake2b(AsRef::<[u8]>::as_ref(&topic_name)).into();
+        let topic_key = Key::Message(MessageAddr::new_topic_addr(entity_addr, topic_name_hash));
+
+        // Check if the topic exists and get the summary.
+        let prev_topic_summary = if let Some(StoredValue::MessageTopic(message_summary)) =
+            self.context.read_gs(&topic_key)?
+        {
+            message_summary
+        } else {
+            return Ok(Err(ApiError::MessageTopicNotRegistered));
+        };
```

Comment:
could be simplified a bit with `let-else` syntax
```suggestion
        let Some(StoredValue::MessageTopic(message_summary)) =
            self.context.read_gs(&topic_key)? else {
            return Ok(Err(ApiError::MessageTopicNotRegistered));
        };
```

Reply:
changed here: https://github.com/casper-network/casper-node/pull/4315/commits/566b3534a29355dd56ec8375f4c42d0a2ca0f24c